plex/player.py

```diff
@@ -0,0 +1,186 @@
+
+        self._itemId = int(itemId)
+
+        for mediaProvider in self._providers.values():
+            importedItems = xbmcmediaimport.getImportedItemsByProvider(mediaProvider)
+            matchingItems = [ importedItem for importedItem in importedItems \
+                if importedItem.getVideoInfoTag() and importedItem.getVideoInfoTag().getUniqueID(PLEX_PROTOCOL) == str(self._itemId) ]
+            if not matchingItems:
+                continue
+
+            if len(matchingItems) > 1:
+                log('multiple items imported from {} match the imported Plex item {} playing from {}' \
+                    .format(mediaProvider2str(mediaProvider), self._itemId, self._file), xbmc.LOGWARNING)
+
+            self._mediaProvider = mediaProvider
+            break
+
+        if self._mediaProvider:
+            # save item
+            plexServer = Server(self._mediaProvider)
+            self._item = Api.getPlexItemDetails(
+                plexServer.PlexServer(),
+                self._itemId,
+                Api.getPlexMediaClassFromMediaType(videoInfoTag.getMediaType())
+            )
+            self._duration = int(self.getTotalTime() * 1000)
+
+            # start playback monitoring thread
+            if not self._playback_monitor or not self._playback_monitor.isAlive():
+                self._playback_monitor = threading.Thread(
+                    target=self._monitor_playback,
+                    name='PLEX:PLAYERMONITOR'
+                )
+                self._playback_monitor.start()
+
+        else:
+            self._reset()
+
+
+    def _syncPlaybackState(self, state=None, time=None):
+        '''Syncs last available state and publishes to PMS'''
+        # either update state or time
+        if not state and not time:
+            return
+
+        # sane check
+        if not self._item:
+            return
+
+        with self._lock:
+            if state:
+                self._last_state['state'] = state
+
+            if time is not None:
+                self._last_state['time'] = int(time)
+
+            # Send update to PMS
+            if self._last_state.get('time') is not None and self._last_state.get('state'):
+                self._item.updateTimeline(
+                    self._last_state['time'],
+                    state=self._last_state['state'],
+                    duration=self._duration
+                )
+
+
+    def _monitor_playback(self):
+        '''Monitor loop that reports current playback state to PMS'''
+        while not self._monitor.abortRequested() and self._item is not None:
+            self._monitor.waitForAbort(REPORTING_INTERVAL)
+            if self._monitor.abortRequested() or not self._item:
+                break
+            self._syncPlaybackState(time=self._getPlayingTime())
+
+
+    def _getPlayingTime(self):
+        '''Gets current xbmc.Player time in miliseconds'''
+        return int(self.getTime() * 1000)
```

Comment:
This conversion to milliseconds appears twice in this code. Maybe add a utility method to `plex.api.Api` (similar to `mediaimporter.emby`).

Reply:
decided to add to lib.utils since it sits in between kodi and plex